src/soroban/assembled_transaction.ts

```diff
@@ -0,0 +1,717 @@
+    parseResultXdr: (xdr: xdr.ScVal) => T;
+  };
+
+export const NULL_ACCOUNT =
+  "GAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAWHF";
+
+export class AssembledTransaction<T> {
+  public raw?: Tx;
+  private simulation?: SorobanRpc.Api.SimulateTransactionResponse;
+  private simulationResult?: SorobanRpc.Api.SimulateHostFunctionResult;
+  private simulationTransactionData?: xdr.SorobanTransactionData;
+  private server: SorobanRpc.Server;
+
+  static ExpiredStateError = class ExpiredStateError extends Error {}
+  static NeedsMoreSignaturesError = class NeedsMoreSignaturesError extends Error {}
+  static WalletDisconnectedError = class WalletDisconnectedError extends Error {}
+  static SendResultOnlyError = class SendResultOnlyError extends Error {}
+  static SendFailedError = class SendFailedError extends Error {}
+  static NoUnsignedNonInvokerAuthEntriesError = class NoUnsignedNonInvokerAuthEntriesError extends Error {}
+
+  static Result = { Ok, Err }
+
+  toJSON() {
+    return JSON.stringify({
+      method: this.options.method,
+      tx: this.raw?.toXDR(),
+      simulationResult: {
+        auth: this.simulationData.result.auth.map((a) => a.toXDR("base64")),
+        retval: this.simulationData.result.retval.toXDR("base64"),
+      },
+      simulationTransactionData:
+        this.simulationData.transactionData.toXDR("base64"),
+    });
+  }
+
+  static fromJSON<T>(
+    options: Omit<AssembledTransactionOptions<T>, "args">,
+    {
+      tx,
+      simulationResult,
+      simulationTransactionData,
+    }: {
+      tx: XDR_BASE64;
+      simulationResult: {
+        auth: XDR_BASE64[];
+        retval: XDR_BASE64;
+      };
+      simulationTransactionData: XDR_BASE64;
+    }
+  ): AssembledTransaction<T> {
+    const txn = new AssembledTransaction(options);
+    txn.raw = TransactionBuilder.fromXDR(tx, options.networkPassphrase) as Tx;
+    txn.simulationResult = {
+      auth: simulationResult.auth.map((a) =>
+        xdr.SorobanAuthorizationEntry.fromXDR(a, "base64")
+      ),
+      retval: xdr.ScVal.fromXDR(simulationResult.retval, "base64"),
+    };
+    txn.simulationTransactionData = xdr.SorobanTransactionData.fromXDR(
+      simulationTransactionData,
+      "base64"
+    );
+    return txn;
+  }
+
+  private constructor(public options: AssembledTransactionOptions<T>) {
+    this.server = new SorobanRpc.Server(this.options.rpcUrl, {
+      allowHttp: this.options.rpcUrl.startsWith("http://"),
+    });
+  }
+
+  static async fromSimulation<T>(
+    options: AssembledTransactionOptions<T>
+  ): Promise<AssembledTransaction<T>> {
+    const tx = new AssembledTransaction(options);
+    const contract = new Contract(options.contractId);
+
+    tx.raw = new TransactionBuilder(await tx.getAccount(), {
+      fee: options.fee?.toString(10) ?? BASE_FEE,
+      networkPassphrase: options.networkPassphrase,
+    })
+      .addOperation(contract.call(options.method, ...(options.args ?? [])))
+      .setTimeout(TimeoutInfinite)
```

Comment:
this was mentioned in the CLI meeting today - just calling out for posterity that we likely want to implement a bound here

Reply:
Yes, I was thinking this could be done in a follow-up PR, after the initial shape of the code lands here. This is not adding anything new to the logic used by `bindings-ts`, it's only exposing it more to people who know better.

Reply:
Sounds good to me 👍. We also need to land on the appropriate timebound defaults.